src/GafferScene/SceneWriter.cpp

```diff
@@ -53,85 +56,101 @@ using namespace GafferScene;
 namespace
 {
 
-void writeLocation( const GafferScene::ScenePlug *scene, const ScenePlug::ScenePath &scenePath, Context *context, IECore::SceneInterface *output, double time )
+tbb::mutex g_writerMutex;
+
+struct LocationWriter
 {
-	ConstCompoundObjectPtr attributes = scene->attributesPlug()->getValue();
-	for( CompoundObject::ObjectMap::const_iterator it = attributes->members().begin(), eIt = attributes->members().end(); it != eIt; it++ )
+	LocationWriter(SceneInterfacePtr output, float time) : m_output( output ), m_time( time )
 	{
-		output->writeAttribute( it->first, it->second.get(), time );
 	}
 
-	if( scenePath.empty() )
+	/// first half of this function can be lock free reading data from ScenePlug
+	/// once all the data has been read then we take a global lock and write
+	/// into the SceneInterface
+	bool operator()( const ScenePlug *scene, const ScenePlug::ScenePath &scenePath )
 	{
-		ConstCompoundObjectPtr globals = scene->globalsPlug()->getValue();
-		output->writeAttribute( "gaffer:globals", globals.get(), time );
-	}
+		ContextPtr context = new Context( *Context::current() );
```

Comment:
This context will be unnecessary once we're computing sets upfront or using `ScenePlug::set()`, but I should mention that this is a very expensive way of making a temporary context. `Context::EditableScope` is the preferred way - it's a good bit quicker and is also designed such that it can be optimised further in the future. The only reason to construct a new context like this is if you need its lifetime to be longer than the scope of the function - typically just in the UI.